ext/AbstractFFTsTestExt.jl

```diff
@@ -52,7 +52,8 @@ const TEST_CASES = (
         )
 
 
-function TestUtils.test_plan(P::AbstractFFTs.Plan, x::AbstractArray, x_transformed::AbstractArray; inplace_plan=false, copy_input=false)
+function TestUtils.test_plan(P::AbstractFFTs.Plan, x::AbstractArray, x_transformed::AbstractArray;
+                             inplace_plan=false, copy_input=false, test_wrappers=true)
```

Comment:
Out of curiosity, why do we have `copy_input`? Couldn't we always copy the input? Or at least always copy when the plan is an in-place plan?

(It would also be good to deduce automatically whether a plan is in-place or not, in the worst case we could check whether `a = copy(x)` and `P * a` are aliased or modify the content of `a`.

Reply:
This is because FFTW in some particular cases mutates the input even with an OOP plans (which has caused users some surprises). This is referenced in the docstring: https://github.com/JuliaMath/AbstractFFTs.jl/blob/a67bf157619b501b48954893b3a0044d6b8e6d0e/src/TestUtils.jl#L33-L34.

Since alternative OOP plan implementations might never mutate their input, they do not need to provide `copy_input=true` which makes for stronger testing of that property.

Reply:
https://github.com/JuliaMath/AbstractFFTs.jl/issues/101 is a bug, isn't it? So I think the `copy_input = true` option should be removed.

Reply:
It's a behaviour of FFTW that's likely never going to change, see for example https://github.com/JuliaMath/FFTW.jl/issues/158 and https://github.com/JuliaMath/AbstractFFTs.jl/pull/88. Even CUFFT has it too I think, since past me added the option in https://github.com/JuliaGPU/CUDA.jl/pull/2073/files#diff-210fa8aa99b3f38d6045fb971b7ee0733923f8f14db526a29040d47d13b36f41R350 too. So I think we need `copy_input = true` to run meaningful tests in these cases

Reply:
I don't see an argument for why it shouldn't be considered a bug in the linked issues and PRs?

Reply:
What do you propose as the path forward? It's a long-standing behaviour of the FFTW C library itself for a couple decades. Without fixing that behaviour we need the `copy_input` option . While the behaviour is unexpected (which is why `copy_input=false` is the default), practically, we need to support this case in the test set, as the only other way forward would be to somehow fix the C library. The behaviour cannot just be patched over by e.g. copying the input array in the plan application, because for many performance use cases the user may not care about the input array anymore after plan application.

It ought to be clearly documented both in `AbstractFFTs` (#88) and by downstream plan backends as a potential behaviour of some plans, which is a separate issue, but I don't see what other path forwards we have for this test set?

Reply:
No, there's no need to fix or change anything in the FFTW library. FFTW.jl just has to copy the input array in this case: https://github.com/JuliaMath/FFTW.jl/issues/158#issuecomment-672244877

Reply:
I won't have the bandwidth to push forward that change forward in `FFTW` unfortunately. (My goal is ultimately just to fix the still-incorrect differentiation of `rfft` plans in `Zygote`.) Also, I think there might be pushback by others who believe the current behaviour should just be documented. Because if one allocates `y` and `x` and has the workflow
```julia
x = something
mul!(y, p, x);  
x = something else
mul!(y, p, x)
x = something else
```
then this code would already be safe and would now go from non-allocating to allocating. And even if it's a very non-ideal situation, presumably there are many reliances of this behaviour in the wild in a library as established as this one.

Reply:
I don't think that the mere presence of the `copy_input` argument, which is `false` by default, should make this PR blocked on handling an independent issue with FFTW.jl / CUFFT.jl that might take on the order of years to be addressed.

Reply:
I just noticed that there's already opposition to creating a copy: https://github.com/JuliaMath/FFTW.jl/issues/158#issuecomment-944397162. So we need to find a solution to this PR that doesn't require us tackling this issue -- is the `copy_input` option (which is `false` by default, only those backends with this mutating behaviour must explicitly set it `true`) acceptable to you? Without it, there is no way to run the adjoint plan tests on downstream trait implementations (which was the original purpose of these test utilities).